### PR TITLE
Added studyprogramme change dialog

### PR DESCRIPTION
**Pull request by anttiviljakainen**

### CODE REVIEW

File: framework/src/main/java/fi/otavanopisto/pyramus/koski/KoskiClient.java [hunk 1]
```diff
@@ -152,42 +141,63 @@ public Oppija findOppijaByOid(String oppijaOid) throws Exception {
...
+          String emsg = String.format("Could not invalidate student oid because source system is not Pyramus (Person %d).", person.getId());
+          logger.log(Level.WARNING, emsg);
+          koskiPersonLogDAO.create(person, KoskiPersonState.UNKNOWN_FAILURE, new Date(), emsg);
+          return false;
         }
       }
+      
+      return updatePersonToKoski(oppija, person, oppijaOid);
     }
+
```
Comment (pkukkon) [inline framework/src/main/java/fi/otavanopisto/pyramus/koski/KoskiClient.java]: Should this be logged as an error, too?

File: framework/src/main/java/fi/otavanopisto/pyramus/koski/KoskiStudentHandler.java [hunk 2]
```diff
@@ -172,7 +172,8 @@ protected void removeInternetixOid(KoskiStudyProgrammeHandler handler, Student s
     
     ObjectMapper mapper = new ObjectMapper();
     try {
-      userVariableDAO.setUserVariable(student, KOSKI_INTERNETIX_STUDYPERMISSION_ID, mapper.writeValueAsString(oids));
+      String variableValue = CollectionUtils.isNotEmpty(oids) ? mapper.writeValueAsString(oids) : "";
```
Comment (pkukkon) [inline framework/src/main/java/fi/otavanopisto/pyramus/koski/KoskiStudentHandler.java]: Does saveOrValidateInternetixOid in this class need the same check?

File: pyramus/src/main/java/fi/otavanopisto/pyramus/json/students/ChangeStudentStudyProgrammeJSONRequestController.java [hunk 3]
```diff
@@ -0,0 +1,58 @@
...
+      client.invalidateAllStudentOIDs(student);
+    } catch (Exception ex) {
+      logger.log(Level.SEVERE, String.format("Invalidation of study permits for student %d failed", studentId), ex);
+    }
+    
+    studentDAO.updateStudyProgramme(student, studyProgramme);
+  }
+
+  public UserRole[] getAllowedRoles() {
+    return new UserRole[] { UserRole.MANAGER, UserRole.STUDY_PROGRAMME_LEADER, UserRole.ADMINISTRATOR };
```
Comment (pkukkon) [inline pyramus/src/main/java/fi/otavanopisto/pyramus/json/students/ChangeStudentStudyProgrammeJSONRequestController.java]: Is this something that could now be limited to just admins, or admins and study programme leaders?

File: framework/src/main/java/fi/otavanopisto/pyramus/koski/KoskiClient.java (hunk 1 above)
Comment (anttiviljakainen) [reply framework/src/main/java/fi/otavanopisto/pyramus/koski/KoskiClient.java]: Logging added

File: framework/src/main/java/fi/otavanopisto/pyramus/koski/KoskiStudentHandler.java (hunk 2 above)
Comment (anttiviljakainen) [reply framework/src/main/java/fi/otavanopisto/pyramus/koski/KoskiStudentHandler.java]: Added

**Pull request merged**